docs/source/distributed.rst

```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
     please contact PyTorch Distributed's maintainers.
 
 
-.. _distributed-launch:
+.. _object_collectives:
+
+Object collectives
+------------------
+
+.. warning::
+    Object collectives have a number of serious limitations.  Read further to determine
+    if they are safe to use for your use case.
+
+Object collectives are a set of collective-like operations that work on arbitrary
+python objects, as long as they can be pickled.  There are various collective patterns
+implemented (e.g. broadcast, all_gather, ...) but they each roughly follow this pattern:
+
+1. convert the input object into a pickle (raw bytes), then shove it into a byte tensor
+2. communicate the size of this byte tensor to peers (first collective operation)
+3. allocate appropriately sized tensor to perform the real collective
+4. communicate the object data (second collective operation)
+5. convert raw data back into python (unpickle)
+
+Object collectives often have surprising performance or memory characteristics that lead to
+long runtimes or OOMs, and thus they should be used with caution.  Here are some common issues.
+
+**Asymetric pickle/unpickle time** - Pickling objects can be slow, depending on the number, type and size of the objects.
+When the collective has a fan-in (e.g. gather_object), the recieving rank(s) must unpickle N times more objects than
+the sending rank(s) had to pickle.  This has led to one class of issues where users report nccl watchdog timeouts caused
+by one rank spending a really long time unpickling data.
```

Comment:
Why would the unpickle time lead to watchdog timeout? I thought the collective has finished at that point? Or are you saying that it may impact the next collective because this gathering rank would join late?

Reply:
yea, its probably always an unrelated collective that actually times out, caused by desync, cuased by unpickling